Feature organization repositories

hristinaina opened the pull request:
Closes #68 

- Implemented repositories section in organization page.
- Please thoroughly verify whether access to repositories and repository permissions are working as excpected. I haven't yet tested the part related to pulling and pushing tags for the repository (but I suppose it should still work like before because I haven't changed anything in registry serivce, only few things in access control service). So if you have time, please test that part as well.

hristinaina committed:
feat: Design Repositories section in Organization

hristinaina committed:
fix: Read access_control for private repos in org

hristinaina committed:
feat: Add admin access control

hristinaina committed:
fix: Define default organization in RepoCreate page

hristinaina committed:
fix: Change navbar access for superadmin

hristinaina committed:
fix: In tests read&write permission is not admin permission

magley commented:
> * Please thoroughly verify whether access to repositories and repository permissions are working as excpected. I haven't yet tested the part related to pulling and pushing tags for the repository (but I suppose it should still work like before because I haven't changed anything in registry serivce, only few things in access control service). So if you have time, please test that part as well.

Here's what I tested:

I have users: `user1, user2, user3`

user1 creates org `org1` and adds user2 and user3
user1 creates repo `repo1` within org1, it's public
user1 adds teams:
- team1, user2 is the only member, team2 has read+write permissions for org1/repo1
- team2, user3 is the only member, team3 has read permissions for org1/repo1

i tried `docker push` to `org1/repo1` with user2, it worked (expected behavior)
i tried `docker push` to `org1/repo1` with user3, it failed (expected behavior)
i tried `docker pull` of `org1/repo1` with user2, it worked (expected behavior)
i tried `docker pull` of `org1/repo1` with user3, it worked (expected behavior)

i tried `docker push` and `docker pull` for `org1/repo1` when signed out, it failed (expected* behavior)

(*) Pushing should definitely fail, I don't know if pulling should fail. The project spec is not clear on this, so we have liberty to define our own business logic.

-----

Test 2: The same thing, but `org1` is now private.

The results were the same (this time, pulling should definitely fail if you are signed out)

magley commented:
Here are some things I've noticed that need fixing:

- when creating a repo, the dropdown (select user or organization) won't go away when I click on it the second time, or when I click away. It always stays there and then it covers the description box
- all organization members should be able to see all teams, but if `user2` clicks on a team he is not part of, the "members" and "permissions" will be empty, and the server returns 400. Can you fix this?

- as a signed out user, I tried to pull a public repo and it failed:
```
docker logout localhost:5000
docker pull localhost:5000/user1/python

Using default tag: latest
Error response from daemon: Head "https://localhost:5000/v2/user1/python/manifests/latest": unauthorized: authentication required
```

`user1/python` exists, I have also pushed some data to it earlier (but this doesn't make a difference), it's a public repo. If I log in, then the pull works.

magley commented on client/src/pages/OrgPage.tsx:
```diff
@@ -110,7 +110,7 @@ export const OrganizationPage = () => {
                             <OrgMembers isActive={key === 'members'} org={org} teamId={-1} />
                         </Tab.Pane>
                         <Tab.Pane eventKey="repositories">
-                            <OrgRepositories isActive={key === 'repositories'} />
+                            <OrgRepositories isActive={key === 'repositories'} org={org} amIMemberOfOrg={amMemberOfOrg} />
                         </Tab.Pane>
                         {amMemberOfOrg && <Tab.Pane eventKey="teams">
```
Could you hide the "settings" tab from the organization page if the user isn't the owner? Because that tab is empty when other org members click on it. I think you can just use `org.owner` and the JWT (if any).

hristinaina replied:
Here's the fix: https://github.com/magley/mocker-hub/pull/75/commits/f7e8b7ab4cb1c294480dfb37c9405fafa25efc8c

magley reviewed (changes requested):
See my comments. It's a bit tricky having to deal with this now. I wish we had tests that automate pushing and pulling images/tags but I couldn't find a way to do this since `docker cli` isn't "ours".

hristinaina committed:
fix: hide settings tab of org & show error

hristinaina commented:
> * when creating a repo, the dropdown (select user or organization) won't go away when I click on it the second time, or when I click away. It always stays there and then it covers the description box
> * all organization members should be able to see all teams, but if `user2` clicks on a team he is not part of, the "members" and "permissions" will be empty, and the server returns 400. Can you fix this?

- Regarding the first issue with the dropdown not closing, f I remember well I think this problem existed before I adjusted the component. I tried looking into it and fixing it but unsuccesfully. 
- For the second issue, I implemented access control in a way that if a user is not a member of a team or owner of org, an AccessDenied error is raised. I saw that Docker Hub also doesn't allow access to this, so I decided to allow users only to see the list of teams, and not the rest. I added a preview for that error message in new commit (https://github.com/magley/mocker-hub/pull/75/commits/f7e8b7ab4cb1c294480dfb37c9405fafa25efc8c); please let me know if you think it’s fine now. 
- Regarding the third point, I won’t be home before 9 PM, so I will look into it when I get home. 

If you have some free time, I would really appreciate it if you could take a look into the first issue.

magley commented:
Alright, if DockerHub does it that way, I'm okay with the way things are. I'll look into the first issue in the meantime.

magley commented:
I think I figured out why the dropdown didn't close. You need to add this import:
```
import 'bootstrap/dist/css/bootstrap.min.css';
```

either at the top of that component or the top of `App.tsx`.

Source: https://stackoverflow.com/a/79372847

Vasilijez commented on server/app/api/access_control/access_control_service.py:
So if I understand correctly, based on the shift in our interpretation of access control, the following changes should be performed in `access_control_service.py`.

Update to:
``` py

    # A user must satisfy one of the following conditions:
    # - Be the owner of the repository.
    # - Be the owner of the organization.
    # - Belong to a team that has either 'admin' or 'read_write' permissions.
    # The same conditions apply to the `has_write_access` method.    
    def has_delete_tag_access(self, user_id: int | None, repo_id: int) -> bool:
        return self.has_write_access(user_id, repo_id)  

```
Remove:
``` py
    # This method is introduced to handle even the special case where the organization owner  
    # is eligible to delete a repository as part of the broader organization deletion process.
    def has_delete_repo_access(self, user_id: int | None, repo: Repository) -> bool:
        org_owner_delete = repo.organization is not None and repo.organization.deleting and repo.organization.owner.id == user_id
        return org_owner_delete or self.has_delete_access(user_id, repo.id)

```
Use `has_admin_access` instead of `has_delete_repo_access`.

If you allow, I can make a commit related to this point, as I have made changes locally.

hristinaina replied:
Of course, please feel free to commit any changes you’ve made.

Vasilijez replied:
I’ve committed it, it’s here: https://github.com/magley/mocker-hub/pull/75/commits/be733853f6a15f547e22a382479e39d03e89cc00

Vasilijez commented on server/app/api/repo/repo_controller.py:
```diff
@@ -74,7 +74,7 @@ def get_repo_by_canonical_name(
     result = repo.model_dump()
     result["owner_name"] = repo.owner.username
     result["org_name"] = None if (repo.organization is None) else repo.organization.name
-    result["can_update"] = access_control_service.has_write_access(user_id, repo.id)
+    result["can_update"] = access_control_service.has_admin_access(user_id, repo.id)
     result["can_star"] = access_control_service.has_star_access(user_id, repo.id)
     result["starred"] = repo_service.is_repo_starred_by(repo, user) if (result["can_star"]) else False
```
I tried to delete the repository as a user with `read_write` permissions, and delete tag button was hidden from me.

This is the case as `can_update` was reused for deleting tag also within `RepoTags.tsx`.

hristinaina replied:
Here's the fix: https://github.com/magley/mocker-hub/pull/75/commits/9556bdffda8feb78b74cc2f2d48047b830692ae1

Vasilijez reviewed (changes requested):
I respect the effort involved in this contribution to fine-tune access control.

I tested many combinations considering the previously mentioned changes, and everything seems to work fine, except for the bugs previously spotted by @magley.

____

Can you please fix this? It should be consistent with the `RepoPreview` description. You can either revert it to how it was before horizontal alignment, or you can update both.
<img width="1919" height="871" alt="image" src="https://github.com/user-attachments/assets/169651f6-0f28-474a-88b8-4345dc25ba44" />

Vasilijez commented:
> * as a signed out user, I tried to pull a public repo and it failed:
> 
> ```
> docker logout localhost:5000
> docker pull localhost:5000/user1/python
> 
> Using default tag: latest
> Error response from daemon: Head "https://localhost:5000/v2/user1/python/manifests/latest": unauthorized: authentication required
> ```
> 
> `user1/python` exists, I have also pushed some data to it earlier (but this doesn't make a difference), it's a public repo. If I log in, then the pull works.

It seems that the issue arises because our backend does not handle the case when the client doesn't forward the auth header. In that case, we simply throw an `HTTPException`.

Vasilijez committed:
fix: Retune access control for `repo` and `tag` deleting

hristinaina committed:
fix: Close dropdown & Change overview of repo and org

hristinaina committed:
fix: Delete tag is controlled by write access

hristinaina commented:
> It seems that the issue arises because our backend does not handle the case when the client doesn't forward the auth header. In that case, we simply throw an `HTTPException`.

If you think we should change how this works, would one of you be able to commit the update? I haven’t worked with that section before, so I’m not really familiar with how it functions. As far as I'm concerned, it can stay as it is now - especially if the project specification didn't mention anything explicitly.

hristinaina committed:
fix: Change dropdown to react select in RepoCreate page

New issue was that by adding "import 'bootstrap/dist/css/bootstrap.min.css';" dropdown on navbar for user profile stopped working. So had to find a new solution.

magley commented:
> If you think we should change how this works, would one of you be able to commit the update?

@Vasilijez and I have come up with a solution for this, he will push it soon enough when he finishes more testing.

Vasilijez committed:
fix: Allow anonymous user to `pull` image

Co-Authored-By: magley <60552482+magley@users.noreply.github.com>

Vasilijez commented:
> @Vasilijez and I have come up with a solution for this, he will push it soon enough when he finishes more testing.

All tests I performed have passed, so everything appears to be fine.
